datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,336 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Distribution partitioner class that works with Hugging Face Datasets."""
+
+
+from collections import Counter
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class DistributionPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on a distribution.
+
+    Inspired from implementations of Li et al.
+    Federated Optimization in Heterogeneous Networks (2020)
+    https://arxiv.org/abs/1812.06127.
+
+    Parameters
+    ----------
+    distribution_array : Union[NDArrayInt, NDArrayFloat]
```

Comment:
Here, we should have either a purely verbal or, if manageable, index-based explanation of what `distribution_array[label_id, index]` means and where such a number of samples will be assigned.

Reply:
Good point, added!